azurelinuxagent/ga/exthandlers.py

```diff
@@ -310,11 +311,74 @@ def handle_ext_handlers(self, etag=None):
                 logger.info("Extension handling is on hold")
                 return
 
+        wait_until = datetime.datetime.utcnow() + datetime.timedelta(minutes=DEFAULT_EXT_TIMEOUT_MINUTES)
+        prev_handler = None
+        ext_seq_required = self.is_ext_seq_required()
+
         self.ext_handlers.extHandlers.sort(key=operator.methodcaller('sort_key'))
         for ext_handler in self.ext_handlers.extHandlers:
-            # TODO: handle install in sequence, enable in parallel
+            if ext_seq_required:
+                # Make sure that the previous extension is handled.
+                # If handled successfully, proceed with the current handler.
+                # Otherwise, skip the rest of them.
+                dep_level = ext_handler.sort_key()
+                if dep_level >= 0:
+                    if self.wait_for_prev_handler_completion(prev_handler, wait_until):
+                        prev_handler = ext_handler
+                    else:
+                        break
+
             self.handle_ext_handler(ext_handler, etag)
-    
+
+    def is_ext_seq_required(self):
+        '''
+        If all the extensions have same dependency level
+        It is considered as extension sequencing is not required
+        '''
+        dep_level = None
+        for ext_handler in self.ext_handlers.extHandlers:
+            if dep_level is None:
+                dep_level = ext_handler.sort_key()
+                continue
+
+            if dep_level != ext_handler.sort_key():
+                return True
+
+        return False
+
+    def wait_for_prev_handler_completion(self, prev_handler, wait_until):
+        '''
+        Check the status of the previous extension handled.
+        Wait until it has a terminal state or times out.
+        Return True if it is handled successfully. False if not.
+        '''
+        # No need to wait on anything for the very first extension
+        if prev_handler == None:
+            return True
+
+        handler_i = ExtHandlerInstance(prev_handler, self.protocol)
+        for ext in prev_handler.properties.extensions:
+            ext_completed, status = handler_i.is_ext_handling_complete(ext)
+
+            # Keep polling for the extension status until it becomes success or times out
+            while not ext_completed and datetime.datetime.utcnow() <= wait_until:
+                time.sleep(5)
+                ext_completed, status = handler_i.is_ext_handling_complete(ext)
+
+            # In case of timeout or terminal error state, we log it and return false
+            # so that the extensions waiting on this one can be skipped processing
+            if not ext_completed or status != EXTENSION_STATUS_SUCCESS:
+                msg = "Extension {0} was timedout or status was: {1}".format(ext.name, status)
```

Comment:
Suggest rephrasing this message to something similar to "Extension {0} did not reach a terminal state within the allowed timeout. Last status was {1}"

Reply:
Disagree. If the extension completes within the allotted time but returns an error, you still wind up emitting this message. Your suggestion makes that error sound like a timeout. The right thing to do is to cleanly separate the "returned an error" condition from the "did not complete within the allotted time limit" condition, emitting a clear (and distinct) message for each of those two cases.

Reply:
@jasonzio , is your concern about the error message or the functionality itself when there is a failure?
As per the last discussion with the VM agent folks, if an extension timedout/failed, we should stop processing the rest of the extensions, when the extension sequencing is required.

Reply:
@jasonzio agree that 2 separate messages would be better